### PR TITLE
refactor(router-lite): better path generation and parameter handling

### PR DESCRIPTION
**Pull request by Sayan751**
# Pull Request

## 📖 Description

This PR is sort of a by-product of the ongoing work of writing the engineering docs for `router-lite`. I tried to fix and improve a couple of things that I noted during the process of writing the docs.  Following is a (non-exhaustive) list of changes in this PR:

- Generation of most matching paths. So far, the path generation using non-path-string (routing-id, routing definition, custom element class etc.) and parameters involved taking the first path in the list of configured paths. This PR changes that and strives to generate a path that consumes most of the given parameters.
- Eager generation of paths and consequently recognized routes is changed. This is related to #1347. The changes introduced in #1467 supported the following syntax: `router.load(routingId, { params: { foo: bar } })`. However, that API is not optimum when dealing with sibling routes. Thus, this PR introduces a BREAKING CHANGE of removing the `params` from the options in favour of `router.load({component, params})`. That way, it scales well for 
  - siblings: `router.load([{component: component1, params: params1}, {component: component2, params: params2}])`, and
  - children: `router.load([{component: component1, params: params1, children: [{component: component2, params: params2}]}])`.
- The support for eager route recognition is also extended from routing-id to routable-VM-instance, custom element etc. (The `component` in the previous point).
- Previously, the `data` in the route configuration can only be a `Record<string, string>`. This PR changes that to `Record<string, unknwon>`, supporting complex data-structures.
- There is a minor bug fix related to querystring. It wasn't working correctly when the navigation is triggered from a non-root routing context.
- Other than these, there are several clean-ups done. Some parts of the APIs are trimmed, as those are no longer relevant.

## 👩‍💻 Reviewer Notes

Note that this is in nature an intermittent PR, as you might see several TODOs in the docs section. I intend to pursue those in my next steps. 

## ⏭ Next Steps

- Change the order of invocation for the `canLoad` and `canUnload` hooks. Currently, for every component, the hooks (global hooks as well as the one in the routable-VM (if any)) are invoked in parallel. Although it makes sense for `load`and `unload` hooks, and maybe to some extent also for the `canUnload` hook, it may cause unexpected results in the scenarios involving `canLoad` hook, especially if the hooks are asynchronous. In my following work, I intend to change the invocation such that the `canLoad` and `canUnload` hooks are called serially.
- Complete writing the engineering doc for `router-lite`.
- Start writing docs for `router-lite`.

### CODE REVIEW

**Review by bigopon — COMMENTED** [review]
> Generation of most matching paths. So far, the path generation using non-path-string (routing-id, routing definition, custom element class etc.) and parameters involved taking the first path in the list of configured paths. This PR changes that and strives to generate a path that consumes most of the given parameters.

I'm not able to get the above part from the changes, can you elaborate it a bit?

**Comment by Sayan751** [comment]
> I'm not able to get the above part from the changes, can you elaborate it a bit?

@bigopon Probably I should have elaborated better in the first place. Let me give it another try. 

The issue and the PR I have linked in the PR description introduced a change where you don't need to use a complete path (example: `foo/123/456`) and rather you can use the route-id and parameters to generate the paths: `load="route:route-id; params.bind:{param1: 123, param2: 345}"`. This way of generating the path, considered only the first configured path previously. This PR changed that. The changed infra should now return a path that consumes the most number of given parameters. Let me also give an example. Let's say you have the following configured route:

```ts
{
  id: 'foo',
  paths: ['foo/:a', 'foo/:a/:b'],
  component: Foo,
}
```

When you try to generate a path for the route-id 'foo' with parameters `{a: 123, b: 456}`, then the generated path will be `foo/123/456` as opposed to `foo/123`, which would have been the case before.

This infra has also been lately extended to support routing definition, custom element class etc. that allows you to do something like this: `router.load({component: Foo, params: {a: 123, b: 456}})`.

That's what I meant by that statement.

File: docs/engineering-notes/router-architecture-notes.md [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+- every route is resolved to a `RouteDefinition`, and
+- the routes are added one after another to the route-recognizer as well as to the navigation-model.
+
+If there is no lazily loaded components then this process is synchronous.
+
+## Generating viewport instruction
+
+This is an internal API used to generate an instance of [`ViewportInstruction`](#viewportinstruction), given a partial `ViewportInstruction` or a route-id or route-definition coupled with parameters.
+
+The idea is to attempt eagerly resolving a path/partial viewport instruction/route definition to a concrete instance of `ViewportInstruction` by finding building a `RecognizedRoute`.
```
[the remark this answers is not in the document]
Comment (Sayan751) [reply docs/engineering-notes/router-architecture-notes.md]: Thank you for noticing that. Pushed changes.

**Comment by bigopon** [comment]
Thanks @Sayan751 , got it, very nice work 💯 🙏

**Pull request merged**